impl/src/main/java/org/ehcache/impl/internal/store/heap/OnHeapStrategy.java

```diff
@@ -0,0 +1,262 @@
+   * @param key key of the mapping
+   * @param valueHolder the mapping
+   * @return new access expiry duration
+   */
+  Duration getAccessDuration(K key, Store.ValueHolder<V> valueHolder);
+
+  /**
+   * Get the new expiry duration as per {@link ExpiryPolicy#getExpiryForUpdate(Object, Supplier, Object)}.
+   *
+   * @param key key of the mapping
+   * @param oldValue the old mapping to be updated
+   * @param newValue the new value for the mapping
+   * @return new access expiry duration
+   */
+  Duration getUpdateDuration(K key, OnHeapValueHolder<V> oldValue, V newValue);
+
+  /**
+   * All purpose strategy. Covers any case that can't be optimized due to the uncertainty of the expiry policy used.
+   *
+   * @param <K> type of the keys stored
+   * @param <V> type of the values stored
+   */
+  class AllStrategy<K, V> implements OnHeapStrategy<K, V> {
+    private final OnHeapStore<K, V> store;
+    private final ExpiryPolicy<? super K, ? super V> expiry;
+    private final TimeSource timeSource;
+
+    public AllStrategy(OnHeapStore<K, V> store, ExpiryPolicy<? super K, ? super V> expiry, TimeSource timeSource) {
+      this.store = store;
+      this.expiry = expiry;
+      this.timeSource = timeSource;
+    }
+
+    @Override
+    public boolean isExpired(OnHeapValueHolder<V> mapping) {
+      return mapping.isExpired(timeSource.getTimeMillis(), TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public void setAccessTimeAndExpiryThenReturnMappingOutsideLock(K key, OnHeapValueHolder<V> valueHolder, long now) {
+      Duration duration = getAccessDuration(key, valueHolder);
+      if (Duration.ZERO.equals(duration)) {
+        // Expires mapping through computeIfPresent
+        store.expireMappingUnderLock(key, valueHolder);
+      } else {
+        valueHolder.accessed(now, duration);
+      }
+    }
+
+    public Duration getAccessDuration(K key, Store.ValueHolder<V> valueHolder) {
+      Duration duration;
+      try {
+        duration = expiry.getExpiryForAccess(key, valueHolder);
+        if (duration != null && duration.isNegative()) {
+          duration = Duration.ZERO;
+        }
+      } catch (RuntimeException re) {
+        LOG.error("Expiry computation caused an exception - Expiry duration will be 0 ", re);
+        duration = Duration.ZERO;
+      }
+      return duration;
+    }
+
+    public Duration getUpdateDuration(K key, OnHeapValueHolder<V> oldValue, V newValue) {
+      Duration duration;
+      try {
+        duration = expiry.getExpiryForUpdate(key, oldValue, newValue);
+        if (duration != null && duration.isNegative()) {
+          duration = Duration.ZERO;
+        }
+      } catch (RuntimeException re) {
+        LOG.error("Expiry computation caused an exception - Expiry duration will be 0 ", re);
+        duration = Duration.ZERO;
+      }
+      return duration;
+    }
+
+    public OnHeapValueHolder<V> setAccessTimeAndExpiryThenReturnMappingUnderLock(K key, OnHeapValueHolder<V> valueHolder, long now,
```

Comment:
This "UnderLock" in this method name is odd in this context. Ut's odd that the name refers to a contract detail related to the calling class.

Reply:
Ok. I found something